pkg/scalers/openstack/keystone_authentication.go

```diff
@@ -0,0 +1,206 @@
+package openstack
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+)
+
+const tokensEndpoint = "/auth/tokens"
+
+// KeystoneAuthMetadata contains all the necessary metadata for Keystone authentication
+type KeystoneAuthMetadata struct {
+	AuthURL    string       `json:"-"`
+	AuthToken  string       `json:"-"`
+	HTTPClient *http.Client `json:"-"`
+	Properties *authProps   `json:"auth"`
+}
+
+type authProps struct {
+	Identity *identityProps `json:"identity"`
+	Scope    *scopeProps    `json:"scope,omitempty"`
+}
+
+type identityProps struct {
+	Methods       []string            `json:"methods"`
+	Password      *passwordProps      `json:"password,omitempty"`
+	AppCredential *appCredentialProps `json:"application_credential,omitempty"`
+}
+
+type passwordProps struct {
+	User *userProps `json:"user"`
+}
+
+type appCredentialProps struct {
+	ID     string `json:"id"`
+	Secret string `json:"secret"`
+}
+
+type scopeProps struct {
+	Project *projectProps `json:"project"`
+}
+
+type userProps struct {
+	ID       string `json:"id"`
+	Password string `json:"password"`
+}
+
+type projectProps struct {
+	ID string `json:"id"`
+}
+
+// GetToken retrieves a token from Keystone
+func (authProps *KeystoneAuthMetadata) GetToken() (string, error) {
+	jsonBody, jsonError := json.Marshal(authProps)
+
+	if jsonError != nil {
+		return "", jsonError
+	}
+
+	body := bytes.NewReader(jsonBody)
+
+	tokenURL, err := url.Parse(authProps.AuthURL)
+
+	if err != nil {
+		return "", fmt.Errorf("the authURL is invalid: %s", err.Error())
+	}
+
+	tokenURL.Path = path.Join(tokenURL.Path, tokensEndpoint)
+
+	getTokenRequest, getTokenRequestError := http.NewRequest("POST", tokenURL.String(), body)
+
+	getTokenRequest.Header.Set("Content-Type", "application/json")
+
+	if getTokenRequestError != nil {
+		return "", getTokenRequestError
+	}
+
+	resp, requestError := authProps.HTTPClient.Do(getTokenRequest)
+
+	if requestError != nil {
+		return "", requestError
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+		authProps.AuthToken = resp.Header["X-Subject-Token"][0]
+		return resp.Header["X-Subject-Token"][0], nil
+	}
+
+	errBody, _ := ioutil.ReadAll(resp.Body)
```

Comment:
nit: since this is an error case, maybe handle the err in case the stream is terminated from the server early.

Reply:
Sorry, I missed this error handling. c6d9f14 added the basic handler.